controller/backing_image_manager_controller.go

```diff
@@ -0,0 +1,1233 @@
+				continue
+			}
+			if info.State == types.BackingImageDownloadStateFailed {
+				continue
+			}
+			pullRequired = false
+			// Use images in default manager as senders only
+			if bim.Spec.Image == defaultImage && info.State == types.BackingImageDownloadStateDownloaded && info.SendingReference < bimtypes.SendingLimit {
+				senderCandidate = bim
+				break
+			}
+		}
+
+		if pullRequired {
+			isEligible, err := c.isEligibleForPulling(currentBIM, biName)
+			if err != nil {
+				return err
+			}
+			if !isEligible {
+				log.Debugf("Current backing image manager is not eligible for pulling")
+				continue
+			}
+			log.Debugf("Start to pull backing image")
+			if _, err := cli.Pull(bi.Name, bi.Spec.ImageURL, bi.Status.UUID); err != nil {
+				if types.ErrorAlreadyExists(err) {
+					log.Debugf("Backing image already exists, no need to pull it again")
+					continue
+				}
+				return err
+			}
+			log.Debugf("Pulling backing image")
+			c.eventRecorder.Eventf(currentBIM, v1.EventTypeNormal, EventReasonPulling, "Pulling backing image %v in disk %v on node %v", bi.Name, currentBIM.Spec.DiskUUID, currentBIM.Spec.NodeID)
+			continue
+		}
+		if senderCandidate != nil {
+			log.WithFields(logrus.Fields{"fromHost": senderCandidate.Status.IP, "toHost": currentBIM.Status.IP, "size": bi.Status.Size}).Debugf("Start to sync backing image")
+			if _, err := cli.Sync(biName, bi.Spec.ImageURL, bi.Status.UUID, senderCandidate.Status.IP, currentBIM.Status.IP, bi.Status.Size); err != nil {
+				if types.ErrorAlreadyExists(err) {
+					log.WithFields(logrus.Fields{"fromHost": senderCandidate.Status.IP, "toHost": currentBIM.Status.IP, "size": bi.Status.Size}).Debugf("Backing image already exists, no need to sync from others")
+					continue
+				}
+				return err
+			}
+			log.WithFields(logrus.Fields{"fromHost": senderCandidate.Status.IP, "toHost": currentBIM.Status.IP, "size": bi.Status.Size}).Debugf("Syncing backing image")
+			c.eventRecorder.Eventf(currentBIM, v1.EventTypeNormal, EventReasonSyncing, "Syncing backing image %v in disk %v on node %v from %v(%v)", bi.Name, currentBIM.Spec.DiskUUID, currentBIM.Spec.NodeID, senderCandidate.Name, senderCandidate.Status.IP)
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (c *BackingImageManagerController) isEligibleForPulling(currentBIM *longhorn.BackingImageManager, biName string) (bool, error) {
+	defaultBIMs, err := c.ds.ListDefaultBackingImageManagers()
+	if err != nil {
+		return false, err
+	}
+	readyCandidateNames := []string{}
+	for bimName, bim := range defaultBIMs {
+		if _, exists := bim.Spec.BackingImages[biName]; !exists {
+			continue
+		}
+		isDownOrDeleted, err := c.ds.IsNodeDownOrDeleted(bim.Spec.NodeID)
+		if err != nil {
+			return false, err
+		}
+		if isDownOrDeleted {
+			continue
+		}
+		readyCandidateNames = append(readyCandidateNames, bimName)
+	}
+	sort.Strings(readyCandidateNames)
+	if len(readyCandidateNames) > 0 && currentBIM.Name == readyCandidateNames[0] {
+		return true, nil
+	}
```

Comment:
To distribute the downloads for each BI among BIMs while still keep the logic simple and make sure that the only 1 BIM downloads the BI. Can we try this:
```
func (c *BackingImageManagerController) isEligibleForPulling(currentBIM *longhorn.BackingImageManager, biName string) (bool, error) {
	defaultBIMs, err := c.ds.ListDefaultBackingImageManagers()
	if err != nil {
		return false, err
	}
	candidateChecksums := []string{}
	candidates := map[string]string{}
	for bimName, bim := range defaultBIMs {
		if _, exists := bim.Spec.BackingImages[biName]; !exists {
			continue
		}
		if bim.Status.CurrentState != types.BackingImageManagerStateRunning {
			continue
		}
		candidates[util.GetStringChecksum(biName+bimName)] = bimName
		candidateChecksums = append(candidateChecksums, bimName)
	}
	sort.Strings(candidateChecksums)
	if len(candidateChecksums) > 0 && currentBIM.Name == candidates[candidateChecksums[0]] {
		return true, nil
	}
	return false, nil
}
```

Reply:
Good idea. Will apply this.